src/render_text.c

```diff
@@ -0,0 +1,598 @@
+                ScrollWindow(printer->template.window,
+                    0,
+                    printer->scrollDistance,
+                    (u8)(printer->template.bgColor | (printer->template.bgColor << 4)));
+                printer->scrollDistance = 0;
+            } else {
+                ScrollWindow(printer->template.window,
+                    0,
+                    4,
+                    (u8)(printer->template.bgColor | (printer->template.bgColor << 4)));
+
+                printer->scrollDistance -= 4;
+            }
+
+            CopyWindowToVram(printer->template.window);
+        } else {
+            printer->state = 0;
+        }
+
+        return RENDER_UPDATE;
+    case 5:
+        printer->state = 0;
+        return RENDER_UPDATE;
+    case 6:
+        if (printer->delayCounter != 0) {
+            printer->delayCounter--;
+        } else {
+            printer->state = 0;
+        }
+
+        return RENDER_UPDATE;
+
+    case 7:
+        if (TextPrinterWaitWithDownArrow(printer)) {
+            TextPrinterClearDownArrow(printer);
+            FillWindowPixelBuffer(printer->template.window, printer->template.bgColor);
+            printer->template.currentX = printer->template.x;
+            printer->template.currentY = printer->template.y;
+            printer->state             = 0;
+        }
+        return RENDER_UPDATE;
+    case 8:
+        if (TextPrinterWaitWithDownArrow(printer)) {
+            TextPrinterClearDownArrow(printer);
+            printer->scrollDistance    = printer->template.lineSpacing + GetFontAttribute(printer->template.fontId, 1);
+            printer->template.currentX = printer->template.x;
+            printer->state             = 4;
+        }
+        return RENDER_UPDATE;
+    }
+
+    return RENDER_FINISH;
+}
+
+void sub_020027F0(int flag) {
+    unk00 = flag;
+}
+
+static void TextPrinterInitDownArrowCounters(TextPrinter *printer) {
+    TextPrinterSubStruct *subStruct = (TextPrinterSubStruct *)(&printer->subStructFields);
+
+    if (sTextFlags.autoScroll) {
+        subStruct->autoScrollDelay = 0;
+    } else {
+        subStruct->downArrowYPosIdx = 0;
+        subStruct->downArrowDelay   = 0;
+    }
+    sub_0200EB68(printer->template.window, unk00);
+}
+
+static void TextPrinterDrawDownArrow(TextPrinter *printer) {
+    TextPrinterSubStruct *subStruct = (TextPrinterSubStruct *)(&printer->subStructFields);
+
+    if (sTextFlags.autoScroll || sTextFlags.useAlternateDownArrow) {
+        return;
+    }
+
+    if (subStruct->downArrowDelay != 0) {
+        subStruct->downArrowDelay--;
+        return;
+    }
+
+    u8 bg_id  = GetWindowBgId(printer->template.window);
+    u8 x      = GetWindowX(printer->template.window);
+    u8 y      = GetWindowY(printer->template.window);
+    u8 width  = GetWindowWidth(printer->template.window);
+    u8 height = GetWindowHeight(printer->template.window) - 2;
+    u16 r6    = unk00;
+
+    FillBgTilemapRect(printer->template.window->bgConfig,
```

Comment:
any reason why this function doesn't just take (BgConfig, BgId, u16, u8, u8, etc)? Seems like that's the case rather than casting explicitly

Reply:
this was ported from diamond without modifications